docs-sphinx/user-guide/how-to-use-header-only-layoutbuilder.md

```diff
@@ -0,0 +1,267 @@
+===============================================
+
+```{code-cell}
+:tags: [hide-cell]
+
+// Make Awkward headers available in this notebook
+#pragma cling add_include_path("../../src/awkward/_v2/cpp-headers")
+```
+
+What is header-only Layout Builder?
+-----------------------------------
+
+The header-only Layout Builder consists of a set of compile-time, templated, static C++ classes, implemented entirely in header file which can be dropped into any external project, and easily separable from the rest of the Awkward C++ codebase.
+
+The Layout Builder namespace consists of the following 14 types of Layout Builders:
+
+* `Numpy`
+* `ListOffset`
+* `List`
+* `Empty`
+* `EmptyRecord`
+* `Record`
+* `Tuple`
+* `Regular`
+* `Indexed`
+* `IndexedOption`
+* `Unmasked`
+* `ByteMasked`
+* `BitMasked`
+* `Union`
+
+All Builders except `Numpy` and `Empty` can take any other Builder as template parameters.
+These Builders are sufficient to build every type of Awkward Array.
+
+Why header-only Layout Builder?
+-------------------------------
+
+The users can directly include `LayoutBuilder.h` in their compilation, rather than linking against platform-specific libraries or worrying about native dependencies. This makes the integration of Awkward Arrays into other projects easier and more portable.
+
+The code is minimal, it does not include all of the code needed to use Awkward Arrays in Python, nor does it include references to Python or pybind11. The C++ users can use it to make arrays and then copy them to Python without any specialised data types - only raw buffers, strings, and integers.
+
+How to use Layout Builders?
+-----------------------------
+
+The following cpp-headers which reside in this [directory](https://github.com/scikit-hep/awkward/tree/main/src/awkward/_v2/cpp-headers/awkward) are needed to use Layout Builders.
+
+1. BuilderOptions.h
+2. GrowableBuffer.h
+3. LayoutBuilder.h
+4. utils.h
+
+Three phases of using Layout Builder
+------------------------------------
+
+1. <b>Constructing a Layout Builder:</b> from variadic templates. (It is implicit template instantiation).
+2. <b>Filling the Layout Builder:</b> while repeatedly walking over the raw pointers within the LayoutBuilder.
+3. <b>Taking the data out to user allocated buffers:</b> Then user can pass them to Python if they want.
+
+Example
+-------
+First, include the LayoutBuilder header file:
+
+```{code-cell}
+#include "awkward/LayoutBuilder.h"
+```
+
+Below is an example for creating RecordArray with NumpyArray and ListOffsetArray as fields.
+
+Type alias can be used for each builder class.
+
+```{code-cell}
+template<class PRIMITIVE>
+using NumpyBuilder = awkward::LayoutBuilder::Numpy<PRIMITIVE>;
+
+template<class PRIMITIVE, class BUILDER>
+using ListOffsetBuilder = awkward::LayoutBuilder::ListOffset<PRIMITIVE, BUILDER>;
+```
+
+If multiple Record Builders are used in a Builder, then a user-defined `field_map` has to be provided for each Record Builder.
```

Comment:
@ManasviGoyal I haven't looked closely at `LayoutBuilder` yet: could you elaborate here to show how to use single RecordBuilders without a user-defined `field_map`, if that is possible?

Reply:
We can't use the RecordBuilders without a user-defined field_map for now since we are limited to C++14 and can't send strings as template parameters in Record Fields. Although we have two methods to set the field names. Will it be okay in terms of the code cell, if I add the alternative method in a separate cell right after the first method?

Please let me know when you are done with the changes. I`ll address some of the comments by Ianna and make some other changes. Thanks.

Reply:
It is allowed from a c++ standpoint, but what would the builder do if you called both methods, would it error? If so, then you could create a new builder to demonstrate it.

Reply:
I'm finished for now, besides fixing the CI, so my changes shouldn't affect you.

Reply:
It doesn't throw any error, so I can just put it in a separate code cell.

Thanks. I will add my changes.